File: ruby solutions/balanced_binary_tree.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {Boolean}
def is_balanced(root)
    H(root) >= 0
end

def H(root)
    return 0 if !root
    l, r = H(root.left), H(root.right)
    return -1 if l < 0 or r < 0 or (l-r).abs > 1
    [l, r].max + 1
end

File: python solutions/balanced_binary_tree.py
# Definition for a binary tree node.
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        return (self.Height(root) >= 0)
    def Height(self, root) -> int:
        if root is None: return 0
        l, r = self.Height(root.left), self.Height(root.right)
        if l < 0 or r < 0 or abs(l - r) > 1: return -1
        return max(l, r) + 1